text/2394-async_await.md

```diff
@@ -0,0 +1,582 @@
+
+Async/await 和 futures 通常是非同步和並行的強大抽象，並且可能在非同步 IO 空間之外有其他應用。我們今天遇到的案例通常與非同步 IO 相關，但透過引入一等公民語法和 libstd 的支援，我們相信更多的 async 和 await 案例也會蓬勃發展，而這些案例不直接與非同步 IO 相關。
+
+# 教學式解說
+[教學式解說]: #guide-level-explanation
+
+## 非同步函式
+
+函式可以添加 `async` 關鍵字，使它們成為「非同步函式」：
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+非同步函式的工作方式與普通函式不同。呼叫非同步函式時，它不會立即進入主體。相反，它執行實作 `Future` 特徵的匿名型別。當輪詢該 future 時，該函式被執行到它內部的下一個 `await` 或返回點（請參閱接下來 await 語法部分）。
+
+非同步函式是延遲計算的一種 - 在您開始輪詢函式返回的 future 之前，函式本體中沒有任何內容被實際執行。例如：
+
+```rust
+async fn print_async() {
+     println!("Hello from print_async")
+}
+
+fn main() {
+     let future = print_async();
+     println!("Hello from main");
+     futures::executor::block_on(future);
+}
+```
+
+這將在印出 `"Hello from main"` 之前印出 `"Hello from print_async"`。
+
+`async fn foo(args..) -> T` 是 `fn(args..) -> impl Future<Output = T>` 型別的函式。返回型別是編譯器產生的匿名型別。
+
+### `async ||` closures
+
+除了函式，非同步也可以應用於 closure。與非同步函式一樣，非同步 closure 的返回型別為 `impl Future<Output = T>`，而不是 `T`。當您呼叫該 closure 時，它會立即返回一個 future 而不執行任何內容（就像非同步函式一樣）。
+
+```rust
+fn main() {
+    let closure = async || {
+         println!("Hello from async closure.");
+    };
+    println!("Hello from main");
+    let future = closure();
+    println!("Hello from main again");
+    futures::block_on(future);
+}
+```
+
+這將在印出 `"Hello from async closure."` 之前印出兩個 `"Hello from main"`。
+
+`async` closure 可以用 `move` 來捕獲它們包覆在 closure 內的變數的所有權。
+
+## `async` 區塊
+
+您可以使用 `async` 區塊直接將 future 創建為表達式：
+
+```rust
+let my_future = async {
+    println!("Hello from an async block");
+};
+```
+
+這種形式幾乎等同於立即呼叫的 `async` closure。即是：
+
+```rust
+async { /* body */ }
+
+// is equivalent to
+
+(async || { /* body */ })()
+```
+
+除了像 `return`、`break` 和 `continue` 這樣的控制流結構不允許在 `body` 中使用（除非它們出現在一個新的控制流上下文中，比如 closure 或 loop）。 `?`-operator 和 early return 如何在非同步區塊中工作尚未確定（請參閱未解決的問題）。
+
+與 `async` closure 一樣，`async` 區塊可以添加 `move`，以捕獲區塊內所包覆的變數的所有權。
+
+## 編譯器內嵌的 `await!`
+
+一個名為 `await!` 的內嵌函式被添加到編譯器。`await!` 可用於「暫停」future 的計算，將控制權交還給呼叫者。`await!` 接受任何實作 `IntoFuture` 的表達式，並計算為此 future 所傳入的泛型型別（如下面範例的 `Output`）之值。
+
+```rust
+// future: impl Future<Output = usize>
+let n = await!(future);
+```
+
+await 的擴展在它接收到的 future 上重複呼叫 `poll`，當它返回 `Poll::Pending` 時產生對函式的控制，並最終在它返回 `Poll::Ready` 時獲取該項目值。
+
+`await!` 只能在非同步函式、closure 或區塊內使用。在該上下文之外使用它是一個錯誤。
+
+（`await!` 是一個內嵌編譯器，為以後確定其確切語法留出空間。可在未解決的問題部分中查看更多訊息。）
+
+# 技術文件式解說
+[技術文件式解說]: #reference-level-explanation
+
+## 關鍵字
+
+`async` 和 `await` 在 2018 版本中都成為關鍵字。
+
+## `async` 函式、closure、區塊的回傳型別
+
+非同步函式的回傳型別是編譯器生成的唯一匿名型別，類似於 closure 的型別。你可以把這種型別想像成一個枚舉，函式的每個「yield point」都有一個變體 —— 它的開頭、await 表達式和每個回傳。每個變體都儲存需要儲存的狀態，以便從該 yield point 恢復控制。
+
+呼叫函式時，此匿名型別以其初始狀態返回，其中包含此函式的所有引數。
+
+### 特徵綁定
+
+匿名回傳型別實作 `Future`，`Item` 為它的回傳型別。輪詢它會推進函數的狀態，當它到達 `await` 時返回 `Pending` ，當它到達 `return` 時返回 `Ready` 。任何在它已經回傳 `Ready` 一次後對其嘗試進行輪詢都會造成恐慌。
+
+匿名回傳型別對 `Unpin` 特徵有一個相反的實作 - 即 `impl !Unpin`。這是因為未來可能有內部引用，這意味著它永遠不需要被移動。
```

Comment:
Original: The anonymous return type has a negative impl for the Unpin trait - **that is impl !Unpin. This is because the future could have internal references which means it needs to never be moved.**

```
這是因為 future 裡面可能有內部引用，亦即永遠都不應該被移動。
```